casper/src/main/scala/coop/rchain/casper/api/BlockAPIImpl.scala

```diff
@@ -0,0 +1,626 @@
+                      .asRight[Error]
+                  case None => s"${result._1.proposeStatus.show}".asLeft[String]
+                }
+              } yield msg
+            // wait for current propose to finish and return result
+            case Some(resultDef) =>
+              for {
+                // this will hang API call until propose is complete, and then return result
+                // TODO cancel this get when connection drops
+                result <- resultDef.get
+                msg = result._2 match {
+                  case Some(block) =>
+                    s"Success! Block ${block.blockHash.toHexString} created and added."
+                      .asRight[Error]
+                  case None => s"${result._1.proposeStatus.show}".asLeft[String]
+                }
+              } yield msg
+          }
+    } yield r
+
+  override def getListeningNameDataResponse(
+      depth: Int,
+      listeningName: Par,
+      maxBlocksLimit: Int
+  ): F[ApiErr[(Seq[DataWithBlockInfo], Int)]] =
+    if (depth > maxBlocksLimit)
+      s"Your request on getListeningName depth $depth exceed the max limit $maxBlocksLimit"
+        .asLeft[(Seq[DataWithBlockInfo], Int)]
+        .pure[F]
+    else
+      for {
+        mainChain           <- getMainChainFromTip(depth)
+        sortedListeningName <- parSortable.sortMatch[F](listeningName).map(_.term)
+        maybeBlocksWithActiveName <- mainChain.toList.traverse { block =>
+                                      getDataWithBlockInfo(sortedListeningName, block)
+                                    }
+        blocksWithActiveName = maybeBlocksWithActiveName.flatten
+      } yield (blocksWithActiveName, blocksWithActiveName.length).asRight
+
+  override def getListeningNameContinuationResponse(
+      depth: Int,
+      listeningNames: Seq[Par],
+      maxBlocksLimit: Int
+  ): F[ApiErr[(Seq[ContinuationsWithBlockInfo], Int)]] =
+    if (depth > maxBlocksLimit)
+      s"Your request on getListeningNameContinuation depth $depth exceed the max limit $maxBlocksLimit"
+        .asLeft[(Seq[ContinuationsWithBlockInfo], Int)]
+        .pure[F]
+    else
+      for {
+        mainChain <- getMainChainFromTip(depth)
+        sortedListeningNames <- listeningNames.toList
+                                 .traverse(parSortable.sortMatch[F](_).map(_.term))
+        maybeBlocksWithActiveName <- mainChain.toList.traverse { block =>
+                                      getContinuationsWithBlockInfo(
+                                        sortedListeningNames,
+                                        block
+                                      )
+                                    }
+        blocksWithActiveName = maybeBlocksWithActiveName.flatten
+      } yield (blocksWithActiveName, blocksWithActiveName.length).asRight
+
+  private def getMainChainFromTip(depth: Int): F[IndexedSeq[BlockMessage]] =
+    for {
+      tipHashes <- estimator
+      tipHash   = tipHashes.head
```

Comment:
This will crush node when executed, as `estimator` is empty, and .head on empty list is called. I'm not sure that this API call even should be here, as we don't have main parents. But maybe it is better to make it aligned with <https://github.com/rchain/rchain/pull/3661/commits/c156fa1581077a2dd27cbbcd2106f4dfd7462e83> @tgrospic wdy think?

Reply:
`getMainChainFromTip` is already removed in dev so this code needs to be sync with latest changes in BlockAPI.
https://github.com/rchain/rchain/blob/33597197cff60dd3a1afe93576569170ce638b47/casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala#L203-L285

Reply:
I [synchronized](https://github.com/rchain/rchain/pull/3662/commits/929cdf2905dd3a67aabac134695d4edfc55c7508) the code with latest `dev`.